jax/interpreters/partial_eval.py

```diff
@@ -772,67 +773,275 @@ def _remat_partial_eval(trace, _, f, tracers, params):
+    for t in unknown_output_tracers: t.recipe = eqn
+    return _zip_knowns(known_output_tracers, unknown_output_tracers, out_unknowns)
 call_partial_eval_rules[remat_call_p] = _remat_partial_eval
 
 def _partition_knowns(pvals, unknowns: Sequence[bool]):
   return ([e for e, unknown in zip(pvals, unknowns) if not unknown],
           [e for e, unknown in zip(pvals, unknowns) if unknown])
 
 def _zip_knowns(known_list, unknown_list, which_unknown: Sequence[bool]):
+  assert len(known_list) + len(unknown_list) == len(which_unknown)
   known_iter, unknown_iter = iter(known_list), iter(unknown_list)
   return [next(unknown_iter) if uk else next(known_iter) for uk in which_unknown]
 
 
+def _partial_eval_jaxpr_custom(
+    jaxpr: Jaxpr, in_unknowns: List[bool], saveable: Callable[..., bool],
+  ) -> Tuple[Jaxpr, Jaxpr, List[bool], List[bool], int]:
+  if jaxpr.constvars: raise NotImplementedError  # TODO
+  env: Dict[Var, Tuple[bool, bool]] = {}
+  residuals: Set[Var] = set()
+
+  def read(x: Atom) -> Tuple[bool, bool]:
+    if type(x) is Var:
+      return env[x]
+    return (False, True)
+
+  def write(unk: bool, inst: bool, v: Var) -> None:
+    assert (unk, inst) != (True, False)
+    env[v] = (unk, inst)
+
+  def ensure_instantiated(inst: bool, x: Atom) -> Atom:
+    if type(x) is Var and not inst:
+      residuals.add(x)
+    return x
+
+  eqns1, eqns2 = [], []
+  write(False, True, unitvar)
+  map(write, in_unknowns, [True] * len(in_unknowns), jaxpr.invars)
+  for eqn in jaxpr.eqns:
+    unks_in, inst_in = unzip2(map(read, eqn.invars))
+    rule = partial_eval_jaxpr_custom_rules.get(eqn.primitive)
+    if rule:
+      eqn1, eqn2, unks_out, inst_out, res = rule(saveable, unks_in, inst_in, eqn)
+      eqn1 and eqns1.append(eqn1); eqn2 and eqns2.append(eqn2)  # type: ignore
+      residuals.update(res)
+      map(write, unks_out, inst_out, eqn.outvars)
+    elif any(unks_in):
+      inputs = map(ensure_instantiated, inst_in, eqn.invars)
+      eqns2.append(new_jaxpr_eqn(inputs, eqn.outvars, eqn.primitive,
+                                 eqn.params, eqn.source_info))
+      map(partial(write, True, True), eqn.outvars)
+    else:
+      eqns1.append(eqn)
+      if saveable(eqn.primitive, *[x.aval for x in eqn.invars], **eqn.params):
+        map(partial(write, False, False), eqn.outvars)
+      else:
+        inputs = map(ensure_instantiated, inst_in, eqn.invars)
+        eqns2.append(new_jaxpr_eqn(inputs, eqn.outvars, eqn.primitive,
+                                   eqn.params, eqn.source_info))
+        map(partial(write, False, True), eqn.outvars)
+  out_unknowns, out_inst = unzip2(map(read, jaxpr.outvars))
+  assert all(type(v) is Var for v in residuals), residuals
+
+  ins1, _ = partition_list(in_unknowns, jaxpr.invars)
+  outs1_, _ = partition_list(out_unknowns, jaxpr.outvars)
+  outs1 = [x for x in outs1_ if x.aval is not abstract_unit]
+  jaxpr1 = Jaxpr((), ins1, [*outs1, *residuals], eqns1)
```

Comment:
Should we sort the residuals (perhaps by names)? Otherwise we might be generating different jaxprs in different processes and that sounds like a recipe for trouble if one's running a distributed computation